### ISSUE TITLE
[new rule] call mockRestore when jest.spyOn is used

### ISSUE THREAD

**Issue by ondrejsevcik** [jest-community/eslint-plugin-jest#977]
When `jest.spyOn` is created, it is best practice to clear it out after the test is done. This new rule should warn when this is not being done.

```js
test('good test', () => {
  const spy = jest.spyOn(obj, 'prop');
  // ...
  spy.mockRestore();
});

test('bad test', () => {
  const spy = jest.spyOn(obj, 'prop');
  // ...
});
```

**Comment by Kreeg** [comment]
Hello!
I think the problem here is that you can do it just in [jest settings](https://jestjs.io/docs/configuration#restoremocks-boolean). So it's a little complicated because of the need for getting and parsing config object to check whether this option enabled or not

**Comment by ondrejsevcik** [comment]
Maybe It could be disabled by default and enabled only on demand.

**Comment by G-Rath** [comment]
The fact that it's possible via a config option doesn't really come into it (other than recommending that you use that option rather than manage mock restoring at the test level) - the actual problem is how hard it is to accurately track `spyOn` mocks, since there is also `jest.restoreAllMocks()` that could be called, and the spy could be assigned outside of a test.

The other problem is that calling `spy.mockRestore` like that isn't actually enough, since if you have that as the last line of your test it'll only get called _when the test passes_ as otherwise an error is thrown. This can be solved by using a `afterEach`, but then it adds to the complexity of trying to implement a rule that can properly track that.

**Comment by G-Rath** [comment]
I'm going to close this because there's not been a lot of activity and I don't think we can reliably track all the possible ways you could handle this to be able to write a useful rule.
